🚟 Add schema batch 2

shivekkhurana opened the pull request:
Adding Malli Specs for the following components"

Quo / Wallet
- network-amount
- network-bridge
- network-routing
- progress-bar
- required-tokens (schema already exists)
- summary-info

Status: ready
Fixes: #18356

shivekkhurana committed:
🚟 Add schema batch 2

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 2105dcc4 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-18696/1/) | 2024-02-01 16:05:32 | ~6 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240201-155903-2105dc-pr18696-tests.log) |

clauxx commented on src/quo/components/wallet/network_amount/view.cljs:
```diff
@@ -5,7 +5,18 @@
+    [:props
+     [:map {:closed true}
+      [:amount [:string]]
+      [:token [:keyword]]
```
in `quo.token`, `token` can be both keyword or string. I guess here it should be the same

clauxx commented on src/quo/components/wallet/network_bridge/view.cljs:
```diff
@@ -24,6 +25,19 @@
+      [:status [:enum :add :loading :locked]]
+      [:amount [:maybe :string]]
+      [:container-style [:maybe :map]]
+      [:on-press [:maybe :fn]]]]
```
This will fail. The `:fn` schema should be used for adding an anonymous fn that would validate the data e.g.
```clj
[:amount [:fn (fn [amount] (> amount 1))]]]
```
Instead you can just use the `fn?` predicate

clauxx commented on src/quo/components/wallet/network_routing/view.cljs:
```diff
@@ -172,6 +173,16 @@
+   [:catn
+    [:props
+     [:map {:closed true}
+      [:networks [:maybe :coll-of :any]]
```
The `:coll-of` should go into its own vector like
```clj
[:networks [:maybe [:coll-of :any]]]

```
But also, to make sure that both collections and vectors can be passed, would be better to replace `:coll-of` with `:sequential`

clauxx commented on src/quo/components/wallet/network_routing/view.cljs:
```diff
@@ -172,6 +173,16 @@
+    [:props
+     [:map {:closed true}
+      [:networks [:maybe :coll-of :any]]
+      [:container-style [:maybe :map]]
```
Wherever you use `:maybe` I think we'd also want a `{:optional true}` to cover the cases when the prop is not passed at all. `:maybe` just covers the case when it's `nil`

clauxx commented on src/quo/components/wallet/progress_bar/view.cljs:
```diff
@@ -2,7 +2,16 @@
+    [react-native.core :as rn]
+    [schema.core :as schema]))
+
+(def ?schema
+  [:=>
+   [:catn
+    [:props
+     [:map {:closed true}
+      [:full-width? [:maybe :boolean]]]]
+    :any]])
```
There's more props that are passed to his component than this (check the styles ns)

clauxx commented on src/quo/components/wallet/summary_info/view.cljs:
```diff
@@ -51,6 +52,18 @@
+   [:catn
+    [:props
+     [:map {:closed true}
+      [:theme [:schema.common/theme]]
+      [:type [:keyword]]
+      [:account-props [:map]]
```
If you're not narrowing down the schema, it's enough to just use it without the extra [] around it e.g. `[:type :keyword]`. Otherwise the schema would be invalid.

clauxx reviewed (commented):
I only just noticed the PR is in draft, but already collected some review comments. Will leave them here, but feel free to resolve them if already addressed.

Also, noticed some of the schemas are invalid. I would suggest to test them as you write to make sure they work. If you have questions on that, you can reach out here or on discord in the #malli channel.

shivekkhurana commented:
@clauxx Thanks for the review. I'm working on this and most issues are already resolved. This specific commit was a first pass that I did with ChatGPT, just to check how feasible it would be.

shivekkhurana committed:
🚈 Fix issues with AI translation

shivekkhurana committed:
🧵 Change amount from react-child to string

shivekkhurana committed:
🧵 Can't just be string, need a `:maybe`

shivekkhurana committed:
⏬ Add maybe and optionals

ilmotta commented on src/quo/components/wallet/progress_bar/view.cljs:
```diff
@@ -11,7 +11,7 @@
      [:map {:closed false}
       [:customization-color {:optional true} [:maybe :schema.common/customization-color]]
-      [:theme :schema.common/theme]
+      [:theme {:optional true} :schema.common/theme]
```
@shivekkhurana, could you help me understand why we need to make this optional? The `view` component always gets the `theme` value injected by `with-theme`. So in which case does the prod code get called without the `theme` key present?

shivekkhurana replied:
When I was writing this, I was assuming that, in case no theme is passed, the default theme is applied.

 So I made it optional. After reading your message, I looked into it and realised that a theme value is expected explicitly. So I'll make this a required field.

ilmotta commented on src/quo/components/wallet/network_amount/view.cljs:
```diff
@@ -5,7 +5,18 @@
+    [:props
+     [:map {:closed true}
+      [:amount {:optional true} [:maybe :string]]
+      [:token {:optional true} [:or keyword? string?]]
```
We haven't reached a conclusion yet about using keywords or functions as predicates, but here it would be better to be consistent and use only keywords or only functions.

shivekkhurana committed:
🧑‍⚖️ Make theme a required prop

shivekkhurana commented:
I'm getting multiple schema errors in CI:
```
[2024-02-05T10:24:15.289Z]     console.error
[2024-02-05T10:24:15.289Z]       2024-02-05T10:24:06.962Z ERROR [schema.core:90] - Failed to instrument function {:schema-id quo.components.wallet.progress-bar.view/view-internal, :error #error {:message ":malli.core/child-error", :data {:type :malli.core/child-error, :message :malli.core/child-error, :data {:type :=>, :properties nil, :children [[:catn [:props [:map {:closed false} [:customization-color {:optional true} [:maybe :schema.common/customization-color]] [:theme :schema.common/theme] [:progressed-value {:optional true} [:maybe :number]] [:full-width? {:optional true} [:maybe :boolean]]]] :any]], :min 2, :max 2}}}, :function #'quo.components.wallet.progress-bar.view/view-internal}```

Checking what's up...

shivekkhurana committed:
🍙 Fix misplaced square brackets that broke spec

ilmotta commented on src/quo/components/wallet/progress_bar/view.cljs:
```diff
@@ -2,7 +2,19 @@
+  [:=>
+   [:catn
+    [:props
+     [:map {:closed false}
```
Map schemas are open by default, thus we can remove the explicit `:closed` option.

shivekkhurana committed:
🎨 Assume default theme and fix tests

- Fixes #18734

shivekkhurana committed:
⬆️ Update schema and rebase

shivekkhurana committed:
🤯 Fix lint

J-Son89 commented on src/test_helpers/component.cljs:
```diff
@@ -53,8 +53,10 @@
   (rtl/render (reagent/as-element component)))
 
 (defn render-with-theme-provider
```
@ilmotta - I believe your suggestion was to update `render` method instead, right??

shivekkhurana replied:
I tried changing the render method but that seems to be incorrect. 
It throws this error:

```
● share wallet addresses › should display the the wallet tab

    Unable to find node on an unmounted component.

      at Function.waitFor [as cljs$core$IFn$_invoke$arity$2] (test_helpers/component.cljs:78:5)
      at cljs$core$IFn$_invoke$arity$2 (test_helpers/component.cljs:75:1)
      at Function.call [as cljs$core$IFn$_invoke$arity$1] (test_helpers/component.cljs:76:17)
      at cljs$core$IFn$_invoke$arity$1 (test_helpers/component.cljs:75:1)
      at Object.call (status_im/contexts/shell/share/wallet/component_spec.cljs:29:6)
```

The approach I implemented: 
- render method with no provider as usual
- render-with-provider method with a default theme is what works 

I don't have details on why adding a provider `render` method fails.

shivekkhurana committed:
🆙 Update progress bar value to be string or int

- It's int when it's 0, and string otherwise
- Update confirmation progress tests to render with quo provider

shivekkhurana committed:
🧪 Update tests

- There is some issue with how :theme prop is drilled in
- Multiple tests are still failing

shivekkhurana committed:
🔨 Fix lint

shivekkhurana committed:
Merge branch 'develop' into 18356-wallet-add-malli-spec-to-quo-wallet-components-batch-2

shivekkhurana committed:
🔨 Fix lint

clauxx commented on src/quo/components/wallet/summary_info/view.cljs:
```diff
@@ -51,6 +52,18 @@
+      [:type [:enum :status-account :saved-account :account :user]]
+      [:account-props {:optional true} [:maybe :map]]
+      [:networks? {:optional true} [:maybe :boolean]]
+      [:values {:optional true} [:maybe :any]]]]]
```
Why `:any`?

shivekkhurana replied:
Fixed it.

clauxx commented on src/quo/components/wallet/network_routing/view.cljs:
```diff
@@ -172,6 +173,16 @@
+   [:catn
+    [:props
+     [:map
+      [:networks {:optional true} [:maybe [:sequential :any]]]
```
If we have all the networks in a collection, the schema could check if it is homogenous over it?

shivekkhurana replied:
Done

shivekkhurana committed:
🔩 Tighten schema

shivekkhurana committed:
Merge branch 'develop' into 18356-wallet-add-malli-spec-to-quo-wallet-components-batch-2

shivekkhurana committed:
⚒️ Fix lint
